keps/0019-package-api-versioning.md

```diff
@@ -2,64 +2,374 @@
+
+Allow operator users to:
+
+- Install packages with a specific _package version_
+- Install packages' *latest* versions
+- Query the available _package versions_ for a _package_
+- Query the available _package versions_ for a _package_ that are compatible with a specific version of Kubernetes
+- Query the available _package versions_ for a _package_ that are compatible with a specific version of the KUDO API
+- Receive operator-related features and bug fixes on operators based on a non-latest _app version_
 
 ### Non-Goals
 
-* Handle operator dependencies on specific Kubernetes versions
+- Explain or define the scenario of multiple KUDO versions with possibly multiple KUDO API versions running on a single Kubernetes cluster
+- Explain or define the planned implementation of operator<->operator dependency graphs (e.g., Kafka operator based on Kafka 2.3.1 requires a Zookeeper operator instance with version above 3.4.10)
+- Explain or define the relationship between KUDO version and KUDO API version
+- Explain or define implementations for transparent and automatic handling of multiple package registries from the KUDO CLI program
+
+## Current State
 
-## Proposal
+An `operator.yaml` file provides versions for all components describing an operator:
 
-A `operator.yaml` file already provides versions for all components describing an operator:
-  * The dependency on KUDO is provided by the `apiVersion` and `kudoVersion` fields
-  * The bundled application version is provided by the `appVersion` field
-  * The version of the operator is provided by the `version` field
-  * The version of Kubernetes is provided by the `kubernetesVersion` field
+- The _KUDO API version_ is specified in the `apiVersion` field
+- The _KUDO version_ is specified in the `kudoVersion` field
+- The _application version_ is specified in the `appVersion` field
+- The _operator version_ is specified in the `version` field
+- The minimum Kubernetes version is specified in the `kubernetesVersion` field
 
 From these versions, currently only the `version` field can be used when installing packages with `kubectl kudo install`. Furthermore, a package repository index only contains `appVersion` and `version` fields in its entries. This limits the possible installation scenarios.
-At least the following installation scenarios should be supported:
-  1. A package repository could contain operators for older `apiVersion` and than the currently released KUDO. A user should be able to install these operators with older versions of KUDO.
-  2. A user should be able to install a specific `appVersion` of an operator
 
-### Support multiple API versions in the package repository
+## Proposals
+
+### With regards to package version and operator version
+
+Two proposals: singular and composite package/operator version. In the case of potential operator "bundles" (operators containing multiple app versions) both would end up "hiding" those app versions behind the package/operator version.
```

Comment:
What does "hiding" mean here? As described above, a package already includes all these versions in `operator.yaml`.